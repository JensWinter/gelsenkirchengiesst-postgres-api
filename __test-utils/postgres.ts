// db.js
import postgres from "postgres";
// This is the local database spawned by supabase
// You should not use your prodction database here. It will inject/remove data
// Therefore this is currently hardcoded to localhost
const url = "postgresql://postgres:postgres@localhost:54322/postgres";

export async function truncateTreesWaterd() {
	const sql = postgres(url);
	await sql`TRUNCATE TABLE trees_watered`;
	sql.end();
}
export async function truncateTreesAdopted() {
	const sql = postgres(url);
	await sql`TRUNCATE TABLE trees_adopted`;
	sql.end();
}

export async function createWateredTrees() {
	const sql = postgres(url);
	await sql`
		INSERT INTO trees_watered (uuid, tree_id, amount, timestamp)
		VALUES
			('test', '_2100294b1f', 1, '2023-01-01 00:00:00'),
			('test', '_2100294b1f', 1, '2023-01-01 00:00:00'),
			('test', '_2100186c08', 1, '2023-01-01 00:00:00'),
			('test', '_2100186c08', 1, '2023-01-01 00:00:00');
			`;
	sql.end();
}
