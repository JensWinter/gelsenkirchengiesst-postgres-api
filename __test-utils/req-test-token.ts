const issuer = process.env.issuer || "";
const client_id = process.env.client_id || "";
const client_secret = process.env.client_secret || "";
const audience = process.env.audience || "";

export async function requestTestToken() {
	const response = await fetch(`${issuer}oauth/token`, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
		},
		body: JSON.stringify({
			client_id,
			client_secret,
			audience,
			grant_type: "client_credentials",
		}),
	});
	if (!response.ok) {
		const json = await response.text();
		throw new Error(`Could not get test token, ${json}`);
	}
	const json = await response.json();
	return json.access_token;
}
