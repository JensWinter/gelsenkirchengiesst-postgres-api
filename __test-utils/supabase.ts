import { createClient } from "@supabase/supabase-js";
import { Database } from "../_types/database";
export const SUPABASE_URL =
	process.env.SUPABASE_URL || "http://localhost:54321";
export const SUPABASE_ANON_KEY = process.env.SUPABASE_ANON_KEY || "";
export const SUPABASE_SERVICE_ROLE_KEY =
	process.env.SUPABASE_SERVICE_ROLE_KEY || "";

export const supabaseAnonClient = createClient<Database>(
	SUPABASE_URL,
	SUPABASE_ANON_KEY
);
export const supabaseServiceRoleClient = createClient<Database>(
	SUPABASE_URL,
	SUPABASE_SERVICE_ROLE_KEY
);
