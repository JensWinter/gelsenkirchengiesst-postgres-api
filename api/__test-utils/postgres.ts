// db.js
import postgres from "postgres";
const url = "postgresql://postgres:postgres@localhost:54322/postgres";

export async function truncateTreesWaterd() {
	const sql = postgres(url);
	await sql`TRUNCATE TABLE trees_watered`;
	sql.end();
}
export async function truncateTreesAdopted() {
	const sql = postgres(url);
	await sql`TRUNCATE TABLE trees_adopted`;
	sql.end();
}
