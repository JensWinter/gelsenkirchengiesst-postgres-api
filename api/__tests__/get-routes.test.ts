import each from "jest-each";
import fetch from "cross-fetch";
import { test, describe, expect } from "@jest/globals";
import handler from "../get/[type]";
import { createTestServer } from "../__test-utils/create-test-server";
import {
	createWateredTrees,
	truncateTreesWaterd,
} from "../__test-utils/postgres";
import { requestTestToken } from "../__test-utils/req-test-token";
// byid ✓
// watered ✓
// all ✓
// treesbyids  ✓
// countbyage ✓
// byage ✓
// wateredandadopted ✓
// lastwatered ✓
//
// the tests below need auth
// adopted  ✓
// istreeadopted ✓
// wateredbyuser ✓

describe("GET routes snapshot tests default responses", () => {
	test("should return 200 on wateredbyuser route authenticated", async () => {
		const token = await requestTestToken();
		const { server, url } = await createTestServer(
			{ type: "wateredbyuser", uuid: "auth0|abc" },
			handler
		);
		const response = await fetch(url, {
			headers: {
				authorization: `Bearer ${token}`,
			},
		});
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("should return 200 on istreeadopted route authenticated", async () => {
		const token = await requestTestToken();
		const { server, url } = await createTestServer(
			{ type: "istreeadopted", id: "_210028b9c8", uuid: "auth0|abc" },
			handler
		);
		const response = await fetch(url, {
			headers: {
				authorization: `Bearer ${token}`,
			},
		});
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});

	test("should return 200 on adopted route authenticated", async () => {
		const token = await requestTestToken();
		const { server, url } = await createTestServer(
			{ type: "adopted", uuid: "auth0|abc" },
			handler
		);
		const response = await fetch(url, {
			headers: {
				authorization: `Bearer ${token}`,
			},
		});
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("should return 200 on lastwatered route", async () => {
		const { server, url } = await createTestServer(
			{ type: "lastwatered", id: "_210028b9c8" },
			handler
		);
		const response = await fetch(url);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("should return 200 on wateredandadopted route", async () => {
		const { server, url } = await createTestServer(
			{ type: "wateredandadopted" },
			handler
		);
		const response = await fetch(url);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("should return 200 on byage route", async () => {
		const { server, url } = await createTestServer(
			{ type: "byage", start: "1800", end: "3000" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});

	test("should return 200 on wateredandadopted", async () => {
		const { server, url } = await createTestServer(
			{ type: "wateredandadopted" },
			handler
		);
		const response = await fetch(url);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("should return 200 on countbyage route", async () => {
		const { server, url } = await createTestServer(
			{ type: "countbyage", start: "1800", end: "3000" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});

	test("Should return 200 on treesbyid route", async () => {
		const { server, url } = await createTestServer(
			{ type: "treesbyids", tree_ids: "_2100294b1f,_210028b9c8" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		const json = await response.json();

		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("should return 200 on trees_watered watered route", async () => {
		await truncateTreesWaterd();
		await createWateredTrees();
		const { server, url } = await createTestServer(
			{ type: "watered" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);

		expect(json).toMatchSnapshot();
		await truncateTreesWaterd();
	});

	test("should return 200 on tree by id route", async () => {
		const { server, url } = await createTestServer(
			{ type: "byid", id: "_2100294b1f" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});
	test("Should return 200 an tree all route", async () => {
		const { server, url } = await createTestServer(
			{ type: "all", limit: "2", offset: "0" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		const json = await response.json();
		expect(response.status).toBe(200);
		expect(json).toMatchSnapshot();
	});

	test("should return 404 on invalid route", async () => {
		const { server, url } = await createTestServer(
			{ type: "invalid" },
			handler
		);
		const response = await fetch(`${url}`);
		server.close();
		expect(response.status).toBe(404);
		expect(response.statusText).toBe("Not Found");
		expect(await response.json()).toMatchSnapshot();
	});
});

each([
	[400, "all", { limit: "abc" }, "due to limit being NaN"],
	[400, "all", { limit: 10000000 }, "due to limit being to large"],
	[400, "all", { offset: "abc" }, "due to offset being NaN"],
	[400, "all", { offset: ["1", "2", "3"] }, "due to offset being not a string"],
]).describe.only(
	"should return %d on route %s",
	(
		statusCode: number,
		type: string,
		overrides: Record<string, unknown>,
		description: string
	) => {
		test(`returns ${statusCode} ${description}`, async () => {
			const { server, url } = await createTestServer(
				{ type, ...overrides },
				handler
			);
			const response = await fetch(`${url}`);
			server.close();
			expect(response.status).toBe(statusCode);
			console.log(await response.json());
		});
	}
);
