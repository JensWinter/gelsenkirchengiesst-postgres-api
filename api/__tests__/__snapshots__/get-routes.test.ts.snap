// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET routes (all, byid, watered) Should return 200 an tree all route 1`] = `
{
  "data": [
    [
      "_2100294b1f",
      13.50295,
      52.64778,
      205,
    ],
    [
      "_2100186c09",
      13.50323,
      52.64582,
      205,
    ],
    [
      "_21002949fc",
      13.50014,
      52.64614,
      205,
    ],
  ],
  "error": null,
  "name": "@technologiestiftung/giessdenkiez-de-postgres-api",
  "url": "/",
  "version": "2.0.0",
}
`;

exports[`GET routes (all, byid, watered) should return 200 on tree by id route 1`] = `
{
  "data": [
    {
      "adopted": null,
      "artbot": "Carpinus betulus",
      "artdtsch": "Hainbuche",
      "baumhoehe": "7",
      "bezirk": "Pankow",
      "caretaker": null,
      "eigentuemer": "Land Berlin",
      "gattung": "CARPINUS",
      "gattungdeutsch": "HAINBUCHE",
      "geom": {
        "coordinates": [
          13.50295,
          52.64778,
        ],
        "type": "Point",
      },
      "gmlid": "00008100:00294b1f",
      "hausnr": null,
      "id": "_2100294b1f",
      "kennzeich": "41374",
      "kronedurch": "2",
      "lat": "13.50295",
      "lng": "52.64778",
      "pflanzjahr": 2019,
      "radolan_days": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        4,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        3,
        5,
        6,
        10,
        29,
        37,
        43,
        12,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        3,
        0,
        0,
        1,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        3,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        16,
        6,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        1,
        2,
        0,
        0,
        0,
        10,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        4,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        1,
        0,
        0,
        0,
        0,
        1,
        0,
        0,
        8,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "radolan_sum": 205,
      "stammumfg": "31",
      "standalter": "3",
      "standortnr": "74/2",
      "strname": "HÃ¶rstenweg",
      "type": null,
      "watered": null,
      "zusatz": null,
    },
  ],
  "error": null,
  "name": "@technologiestiftung/giessdenkiez-de-postgres-api",
  "url": "/",
  "version": "2.0.0",
}
`;

exports[`GET routes (all, byid, watered) should return 200 on trees_watered watered route 1`] = `
{
  "data": [
    {
      "tree_id": "_2100294b1f",
    },
    {
      "tree_id": "_2100294b1f",
    },
    {
      "tree_id": "_2100186c08",
    },
    {
      "tree_id": "_2100186c08",
    },
  ],
  "error": null,
  "name": "@technologiestiftung/giessdenkiez-de-postgres-api",
  "url": "/",
  "version": "2.0.0",
}
`;

exports[`GET routes (all, byid, watered) should return 404 on invalid route 1`] = `
{
  "error": "invalid route invalid",
}
`;
